src/Microsoft.Diagnostics.Monitoring.EventPipe/Triggers/AspNet/AspNetRequestDurationTrigger.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe.Triggers.AspNet
+{
+    internal sealed class AspNetRequestDurationTrigger : AspNetTrigger<AspNetRequestDurationTriggerSettings>
+    {
+        private readonly long _durationTicks;
+
+        //This is adjusted due to rounding errors on event counter timestamp math.
+        private readonly TimeSpan _heartBeatInterval = TimeSpan.FromSeconds(AspNetTriggerSourceConfiguration.DefaultHeartbeatInterval - 1);
+        private SlidingWindow _window;
+        private Dictionary<string, DateTime> _requests = new();
+        private DateTime _lastHeartbeatProcessed = DateTime.MinValue;
+
+        public AspNetRequestDurationTrigger(AspNetRequestDurationTriggerSettings settings) : base(settings)
+        {
+            _durationTicks = Settings.RequestDuration.Ticks;
+            _window = new SlidingWindow(settings.SlidingWindowDuration);
+        }
+
+        protected override bool ActivityStart(DateTime timestamp, string activityId)
+        {
+            _requests.Add(activityId, timestamp);
+
+            return false;
+        }
+
+        protected override bool Heartbeat(DateTime timestamp)
+        {
+            //May get additional heartbeats based on multiple counters or extra intervals. We only
+            //process the data periodically.
+            if (timestamp - _lastHeartbeatProcessed > _heartBeatInterval)
+            {
+                _lastHeartbeatProcessed = timestamp;
+                List<string> requestsToRemove = new();
+                bool trigger = false;
+
+                foreach (KeyValuePair<string, DateTime> request in _requests)
+                {
+                    if ((timestamp - request.Value) >= Settings.RequestDuration)
+                    {
+                        _window.AddDataPoint(timestamp);
+
+                        //We don't want to count the request more than once, since it could still finish later.
+                        //At this point we already deeemed it too slow. We also want to make sure we
+                        //clear the cached requests periodically even if they don't finish.
+                        requestsToRemove.Add(request.Key);
+                        if (_window.Count >= Settings.RequestCount)
```

Comment:
If I understand the algorithm correctly, this could be moved to after the request removal loop because you don't always have to check for each request. The count within the window should never get smaller than the value after the first call to AddDataPoint in the loop because you are adding the same timestamp.

This eliminates the need for the trigger flag and allows you to just `return _window.Count >= Settings.RequestCount`.

Reply:
Sure but note i still want to break early so i don't have to iterate through all the requests if i already met the trigger condition.

Reply:
But you need to process all of them so that the count is accurate in case there are more that are above the threshold, correct?

Reply:
If you didn't do that, then you could have a nearly never-ending set of callback invocations (one on each counter tick) if many requests are not completing within the timeout window.